snow/consensus/snowball/consensus_performance_test.go

```diff
@@ -11,46 +11,87 @@ import (
 	"github.com/ava-labs/avalanchego/utils/sampler"
 )
 
-func TestSnowballOptimized(t *testing.T) {
+// Test that a network running the lower AlphaPreference converges faster than a
+// network running equal Alpha values.
+func TestDualAlphaOptimization(t *testing.T) {
 	require := require.New(t)
 
 	numColors := 10
 	numNodes := 100
 	params := Parameters{
-		K: 20, Alpha: 15, BetaVirtuous: 20, BetaRogue: 30,
+		K:               20,
+		AlphaPreference: 15,
+		AlphaConfidence: 15,
+		BetaVirtuous:    15,
+		BetaRogue:       20,
 	}
 	seed := int64(0)
 
-	nBitwise := Network{}
-	nBitwise.Initialize(params, numColors)
+	singleAlphaNetwork := Network{}
+	singleAlphaNetwork.Initialize(params, numColors)
 
-	nNaive := nBitwise
+	params.AlphaPreference = params.K/2 + 1
+	dualAlphaNetwork := Network{}
+	dualAlphaNetwork.Initialize(params, numColors)
 
 	sampler.Seed(seed)
 	for i := 0; i < numNodes; i++ {
-		nBitwise.AddNode(NewTree)
+		dualAlphaNetwork.AddNode(NewTree)
 	}
 
 	sampler.Seed(seed)
 	for i := 0; i < numNodes; i++ {
-		nNaive.AddNode(NewFlat)
+		singleAlphaNetwork.AddNode(NewTree)
 	}
 
+	// Although this can theoretically fail with a correct implementation, it
+	// shouldn't in practice
+	runNetworksInLockstep(require, seed, &dualAlphaNetwork, &singleAlphaNetwork)
+}
+
+// Test that a network running the snowball tree converges faster than a network
+// running the flat snowball protocol.
+func TestTreeConvergenceOptimization(t *testing.T) {
+	require := require.New(t)
+
+	numColors := 10
+	numNodes := 100
+	params := DefaultParameters
+	seed := int64(0)
+
+	treeNetwork := Network{}
+	treeNetwork.Initialize(params, numColors)
+
+	flatNetwork := treeNetwork
+
+	sampler.Seed(seed)
+	for i := 0; i < numNodes; i++ {
+		treeNetwork.AddNode(NewTree)
+	}
+
+	sampler.Seed(seed)
+	for i := 0; i < numNodes; i++ {
+		flatNetwork.AddNode(NewFlat)
+	}
+
+	// Although this can theoretically fail with a correct implementation, it
+	// shouldn't in practice
+	runNetworksInLockstep(require, seed, &treeNetwork, &flatNetwork)
+}
+
+func runNetworksInLockstep(require *require.Assertions, seed int64, fast *Network, slow *Network) {
 	numRounds := 0
-	for !nBitwise.Finalized() && !nBitwise.Disagreement() && !nNaive.Finalized() && !nNaive.Disagreement() {
+	for !fast.Finalized() && !fast.Disagreement() && !slow.Finalized() && !slow.Disagreement() {
 		sampler.Seed(int64(numRounds) + seed)
-		nBitwise.Round()
+		fast.Round()
 
 		sampler.Seed(int64(numRounds) + seed)
-		nNaive.Round()
+		slow.Round()
 		numRounds++
 	}
 
-	require.False(nBitwise.Disagreement())
-	require.False(nNaive.Disagreement())
-
-	// Although this can theoretically fail with a correct implementation, it
-	// shouldn't in practice
-	require.True(nBitwise.Finalized())
-	require.True(nBitwise.Agreement())
+	require.False(fast.Disagreement())
```

Comment:
should this be: `require.False(slow.Finalized())`? I don't see where we check that slow network converges slower than fast one

Reply:
This is checking that `fast` finalizes "at least as fast" as `slow`. It is allowed for `slow` to be finalized here - but not required either way.